### PR TITLE
HWY-269: Make joining after restart more resilient.

### PR DESCRIPTION
**Pull request by goral09**
https://casperlabs.atlassian.net/browse/HWY-269

This PR fixes a bug where a restarted active node is unable to join the current era b/c no other node in the network knows its last unit.

The scenario during which this bug appeared:
* Five node network starts
* Nodes 1, 2, and 3 are stopped after storing its `own_last_unit` to a file but before sending it out to the network
* Network stalls due to inability to finalize anything
* Nodes 1, 2, and 3 are restarted
* Nodes 1-3 try to synchronize their protocol state up to the known `own_last_unit` but are unable to b/c those units were not sent out before they were stopped
* Network stalls forever

This PR makes a couple of changes that should make the recovery more resilient:
1. request latest protocol state on startup regardless of the moment in an era (vs previously where it would be done only if we joined after the era might have ended)
2. create a `Ping` request upon validator activation (signals to the network that we are up and ready)
3. store last known unit (not only hash) to disk. Validator will wait until its protocol state is fully synchronized – i.e. until we have downloaded all of the units in the panorama of our last known unit and then add the unit itself to its protocol state. After that, `can_vote` will return `true` and we will start voting.

We will now also try removing the unit file when deactivating the validator.

### CODE REVIEW

File: node/src/components/consensus/highway_core/active_validator.rs [hunk 1]
```diff
@@ -124,47 +122,70 @@ impl<C: Context> ActiveValidator<C> {
...
     fn can_vote(&self, state: &State<C>) -> bool {
-        self.own_last_unit
-            .map_or(true, |ref hash| state.has_unit(hash))
+        self.own_last_unit.as_ref().map_or(true, |swunit| {
+            state.has_unit(&swunit.hash())
+                && swunit
+                    .wire_unit()
+                    .panorama
+                    .iter_correct_hashes()
+                    .all(|hash| state.has_unit(hash))
```
Comment (afck) [inline node/src/components/consensus/highway_core/active_validator.rs:187]: That's unnecessary: It is implied by `state.has_unit(&swunit.hash())`.

**Review by afck — COMMENTED** [review]
Looks good to me, but I'd make those two changes.

File: node/src/components/consensus/highway_core/active_validator.rs (hunk 1 above)
Comment (goral09) [reply node/src/components/consensus/highway_core/active_validator.rs:187]: Right, my initial version was only checking for the unit's panorama, not the unit itself.

File: node/src/components/consensus/highway_core/active_validator.rs [hunk 2]
```diff
@@ -124,47 +122,70 @@ impl<C: Context> ActiveValidator<C> {
...
         };
-        let effects = av.schedule_timer(start_time, state);
+        let mut effects = av.schedule_timer(start_time, state);
+        effects.push(av.send_ping(current_time));
+        effects.extend(
+            av.own_last_unit
+                .as_ref()
+                .map(|unit| Effect::NewVertex(ValidVertex(Vertex::Unit(unit.clone()))))
+                .into_iter(),
+        );
```
[the remark this answers is not in the document]
Comment (goral09) [reply node/src/components/consensus/highway_core/active_validator.rs:137]: This was a hack to trigger the addition of our own latest unit to _our_ state. The code, in its current form, does not allow for easily adding the `own_last_unit` to state when its panorama is sync'd. I have to back and rewrite more in this case.

[the remark this answers is not in the document]
Comment (afck) [reply node/src/components/consensus/highway_core/active_validator.rs:137]: Right, makes sense, I guess. :weary: 
Would be good to add a TODO then.

File: node/src/components/consensus/highway_core/active_validator.rs (hunk 1 above)
Comment (goral09) [reply node/src/components/consensus/highway_core/active_validator.rs]: Fixed in [`4289f2d` (#1134)](https://github.com/CasperLabs/casper-node/pull/1134/commits/4289f2d02aef82fd23265003ad14867761eafbf9)

File: node/src/components/consensus/highway_core/active_validator.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (goral09) [reply node/src/components/consensus/highway_core/active_validator.rs]: Done in [`4289f2d` (#1134)](https://github.com/CasperLabs/casper-node/pull/1134/commits/4289f2d02aef82fd23265003ad14867761eafbf9).

We won't send our last known unit on validator activation but add it to the state as soon as we have its dependencies satisfied – i.e. when we finish synchronizing with the protocol state up until our latest known unit.

**Review by afck — APPROVED** [review]

**Review by fizyk20 — APPROVED** [review]

**Comment by afck** [comment]
Looks like more tests need to be updated. Otherwise looks good to me.

**Review by afck — APPROVED** [review]

**Comment by goral09** [comment]
bors merge

**Pull request merged**